Only specify `--target` by default for `-Zgcc-ld=lld` on wasm

lqd opened the pull request:
On macOS, it's not yet clear which cases of clang/OS/target/SDK version impact how to find ld/lld/rust-lld. The `--target` argument is not needed on our current targets with a vanilla config to do so, but may be in some cases ? 

That is, things look to be different and more subtle than suggested in https://github.com/rust-lang/rust/issues/97402#issuecomment-1147404520.

Specifying this argument unconditionally currently breaks `-Zgcc-ld=lld` on the 10.7+ targets on x64 macOS. Vanilla configs on x64 and aarch64 don't seem to need it to be able to find `rust-lld`.

This fixes #101653 on macOS (I've tried on x64 and aarch64: vanilla installs with the most recent command line tools available for each of the OS versions I have access to, `run-make/issue-71519` passes).

I didn't expect the previous PRs to fail because of the existing tests, but CI does not actually run those tests yet, which explains the regressions. I was hoping to fix those in this PR but it's more involved (building lld is required for the tests to run, llvm/lld is not built on the test builders but on the dist builders, the dist builders don't run tests). This PR is just to unblock current users on macOS who have reported issues, but a regression could happen in the future by mistake until then.

r? @petrochenkov

lqd committed:
only specify --target by default for -Zgcc-ld=lld on wasm

On macOS, it's not yet clear which cases of clang/OS/target/SDK version impact
how to find ld/lld/rust-lld. The --target is not needed on our current targets with
a vanilla config, but may be in some cases. Specifying it all the time breaks the 10.7+
targets on x64 macOS.

We try to only specify it on macOS if the linker flavors are different,
for possible cases of cross-compilation with `-Zgcc-ld=lld` but the
expectation is that it should be passed manually when needed in these
situations.

tschuett commented:
Maybe you will need the target on macOS for cross-compiling?

lqd commented:
It wasn’t used before, so I’m not sure. And you can still pass it via -Clink-arg I presume ?

Wyvern commented:
Looks like this problem still not fixed.

Another question. Why use `-fuse-ld=lld` generate `clang: error: invalid linker name in argument '-fuse-ld=lld` for now?

lqd commented:
@Wyvern What do you mean not fixed, did you build this PR locally to test ? If so, what error did you get ?

petrochenkov commented:
`10.7` comes from this function in `compiler\rustc_target\src\spec\apple_base.rs`:
```rust
fn macos_default_deployment_target(arch: &str) -> (u32, u32) {
    if arch == "arm64" { (11, 0) } else { (10, 7) }
}
```
, this is the default version that is used if no `MACOSX_DEPLOYMENT_TARGET` is provided by the user.

So setting `MACOSX_DEPLOYMENT_TARGET` that matches your environment should work as a quick workaround for https://github.com/rust-lang/rust/issues/101653.

petrochenkov commented:
@Wyvern 
>Another question. Why use `-fuse-ld=lld` generate `clang: error: invalid linker name in argument '-fuse-ld=lld` for now?

This should only happen if your clang is old (or somehow modified) and doesn't support LLD. In that case it's reasonable that rustc using such clang doesn't support LLD either.

petrochenkov commented:
So if no `MACOSX_DEPLOYMENT_TARGET` is specified, then rustc will always pass target `x86_64-apple-macosx10.7.0` to *LLVM* and apparently that's fine for all the range of supported versions.
However, it's apparently not fine to pass it to *clang* with a different default macosx version.

petrochenkov commented:
>Maybe you will need the target on macOS for cross-compiling?

That's exactly why the `--target` option is passed, it's not necessary if we are not cross-compiling, clang should find the appropriate lld flavor for the host in that case.

There are several possible cases here:
- Not a cross compilation: `--target` is not necessary, the correct target is either hardcoded or used by default.
- Cross compilation:
    - Using a separate toolchain for the specific target (gcc is typically used like this), then `--target` is not necessary because it's hardcoded.
    - Using a toolchain that can switch targets dynamically (e.g. usual clang installed on the host), then `--target` is necessary for that clang to decide what triple to target.

petrochenkov commented:
I think the initial fix should be
```rust
                    if we_are_cross_compiling() {
                        cmd.arg(format!("--target={}", sess.target.llvm_target));
                    }
```
, so the main question is how to implement the `we_are_cross_compiling()` part :D

`sess.host == sess.target` should work in most cases, but I'm not sure it will work for Apple targets specifically due to possible SDK version mismatches between the host and target.

petrochenkov commented:
>so the main question is how to implement the `we_are_cross_compiling()`

`sess.target.lld_flavor != sess.host.lld_flavor`!

tschuett commented:
This could miss the case where I am compiling from OSX to WatchOS. The lld flavour is the same, but I need the target parameter.

petrochenkov commented:
@tschuett 
Sorry, I don't understand.
What do you do in this case when you are compiling without `-Zgcc-lld`? Exactly the same thing should work with `-Zgcc-lld` too.

tschuett commented:
```console
>  cargo check --target x86_64-apple-darwin
```
I am on x86 OSX. The lld flavour is the same, but I would need the target parameter.

Wyvern commented:
Is there a plan to merge this PR into nightly build?

lqd commented:
@Wyvern Yes, we're working on it. Until then, you can also use the `MACOSX_DEPLOYMENT_TARGET` env var and it'll work.  Or a local lld via `-Clink-arg=-fuse-ld=lld` I guess.

Wyvern commented:
`MACOSX_DEPLOYMENT_TARGET=12.0` works, Thx.

bors commented:
:umbrella: The latest upstream changes (presumably #101988) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd commented:
@petrochenkov I've rebased over #101988, and added the simili cross-compiling test for macOS, for the possibly different linker flavors between host and targets. Is this now what you expected ?

I've tested this PR on x64 darwin (without cross compilation shenanigans) and `-Zgcc-ld` will now work. I've also opened a draft PR, waiting on this one, to fix CI so that `needs-rust-lld` tests will be executed to prevent things from regressing again.

@rustbot ready

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2831,11 +2831,31 @@ fn add_gcc_ld_path(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor) {
+                        // Apple targets require (and which would be handled prior to hitting this
+                        // `-Zgcc-ld=lld` codepath anyway), but the expectation is that until then
+                        // this should be manually passed if needed.
+                        let is_targeting_wasm = matches!(flavor, LinkerFlavor::WasmLld(..));
```
Not sure why WASM is special cased here, it is already covered by the linker flavor condition.

lqd replied:
It was less about special casing WASM than having to pick a bool between wasm and darwin, as we have to add `--target` on wasm, and not on darwin. 

So the choice was between `if wasm or different_linker_flavor` and `if !darwin or different_linker_flavor` and the first one seemed clearer to read but I can flip it to the second if you prefer ?

petrochenkov replied:
I mean `different_linker_flavor` should always be true for WASM, so `wasm or different_linker_flavor == different_linker_flavor`.

lqd replied:
Done

petrochenkov commented:
Sorry for the delay, I'm currently taking some break from Rust.
@rustbot author

lqd commented:
> Sorry for the delay, I'm currently taking some break from Rust.

@petrochenkov would you rather we re-roll for another reviewer ?

petrochenkov commented:
@lqd 
No, I reassigned most of my PRs so the remaining ones, including this, should be fine.

lqd commented:
Removed the superfluous bool.

@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 7770b06b06efc4abda75e608e67044a0a344771e has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 7770b06b06efc4abda75e608e67044a0a344771e -->
<!-- homu: {"type":"Approved","sha":"7770b06b06efc4abda75e608e67044a0a344771e","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7770b06b06efc4abda75e608e67044a0a344771e with merge 312c9a37f26d1f0a98fc410109abb961e945b85e...
<!-- homu: {"type":"BuildStarted","head_sha":"7770b06b06efc4abda75e608e67044a0a344771e","merge_sha":"312c9a37f26d1f0a98fc410109abb961e945b85e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3830132969/jobs/6520659284)
Approved by: petrochenkov
Pushing 312c9a37f26d1f0a98fc410109abb961e945b85e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3830132969/jobs/6520659284"},"merge_sha":"312c9a37f26d1f0a98fc410109abb961e945b85e"} -->

rust-timer commented:
Finished benchmarking commit (312c9a37f26d1f0a98fc410109abb961e945b85e): [comparison URL](https://perf.rust-lang.org/compare.html?start=b435960c4cfd3975651c7051be56d7f5d6c201ab&end=312c9a37f26d1f0a98fc410109abb961e945b85e&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b435960c4cfd3975651c7051be56d7f5d6c201ab&end=312c9a37f26d1f0a98fc410109abb961e945b85e&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b435960c4cfd3975651c7051be56d7f5d6c201ab&end=312c9a37f26d1f0a98fc410109abb961e945b85e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.9% | [2.9%, 2.9%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.8% | [2.8%, 2.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.9% | [2.9%, 2.9%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b435960c4cfd3975651c7051be56d7f5d6c201ab&end=312c9a37f26d1f0a98fc410109abb961e945b85e&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9% | [2.9%, 2.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>
